source/libuv/uv_event_loop.c

```diff
@@ -0,0 +1,819 @@
+        aws_linked_list_swap_contents(&impl->cross_thread_data.tasks_to_schedule, &tasks_to_schedule);
+        aws_mutex_unlock(&impl->cross_thread_data.mutex);
+    }
+
+    while (!aws_linked_list_empty(&tasks_to_schedule)) {
+        struct aws_linked_list_node *node = aws_linked_list_pop_front(&tasks_to_schedule);
+        struct aws_task *task = AWS_CONTAINER_OF(node, struct aws_task, node);
+
+        s_schedule_task_impl(impl, task);
+    }
+}
+
+static int s_running_tasks_stop(void *context, struct aws_hash_element *element) {
+
+    (void)context;
+
+    struct task_data *task = element->value;
+
+    /* Stop task and close the handle */
+    uv_timer_stop(&task->timer);
+    uv_close((uv_handle_t *)&task->timer, s_uv_close_timer_no_free);
+
+    return AWS_COMMON_HASH_TABLE_ITER_CONTINUE;
+}
+
+/* Wakes up the event loop and stops it */
+static void s_uv_async_stop_loop(uv_async_t *request) {
+    struct libuv_loop *impl = request->data;
+    assert(impl);
+
+    s_owned(impl)->state = EVENT_THREAD_STATE_STOPPING;
+
+    /* Stop all open subscriptions */
+    struct aws_linked_list_node *it = aws_linked_list_begin(&impl->on_thread_data.open_subscriptions);
+    const struct aws_linked_list_node *end = aws_linked_list_end(&impl->on_thread_data.open_subscriptions);
+    while (it != end) {
+        struct handle_data *handle_data = AWS_CONTAINER_OF(it, struct handle_data, open_subs_node);
+        uv_poll_stop(&handle_data->poll);
+        uv_close((uv_handle_t *)&handle_data->poll, s_uv_close_sub);
+
+        it = it->next;
+    }
+
+    /* Stop all open timers */
+    aws_hash_table_foreach(&impl->on_thread_data.running_tasks, s_running_tasks_stop, NULL);
+
+    uv_close((uv_handle_t *)&impl->stop_async, s_uv_close_handle);
+    uv_close((uv_handle_t *)&impl->schedule_tasks_async, s_uv_close_handle);
+}
+
+static int s_running_tasks_start(void *context, struct aws_hash_element *element) {
+
+    struct task_data *task = element->value;
+    struct libuv_loop *impl = context;
+
+    uv_timer_start(&task->timer, s_uv_task_timer_cb, s_timestamp_to_uv_millis(task->event_loop, task->task->timestamp), 0);
+
+    aws_atomic_fetch_add(&impl->num_open_handles, 1);
+
+    return AWS_COMMON_HASH_TABLE_ITER_CONTINUE;
+}
+
+static int s_run(struct aws_event_loop *event_loop) {
+
+    bool cleanup_async_schedule_tasks = false;
+    bool cleanup_async_stop = false;
+    bool cleanup_polls = false;
+
+    struct libuv_loop *impl = event_loop->impl_data;
+
+    impl->schedule_tasks_async.data = impl;
+    if (uv_async_init(impl->uv_loop, &impl->schedule_tasks_async, s_uv_async_schedule_tasks)) {
+        return AWS_OP_ERR;
+    }
+    cleanup_async_schedule_tasks = true;
+    aws_atomic_fetch_add(&impl->num_open_handles, 1);
+
+    /* Prep the stop async */
+    impl->stop_async.data = impl;
+    if (uv_async_init(impl->uv_loop, &impl->stop_async, s_uv_async_stop_loop)) {
+        goto clean_up;
+    }
+    cleanup_async_stop = true;
```

Comment:
Could we just compare the handles against a sentinel value to know if they were inited, rather than separate flags? I  worry about these getting out of sync if this code is modified.

Reply:
Yeah that's not a bad idea